rust/parsing.rs

```diff
@@ -0,0 +1,945 @@
+                                    let extra_minutes =
+                                        ((extra_hours - extra_full_hours) * 60.0).round();
+                                    let extra_full_minutes: f64 = extra_minutes.trunc();
+                                    duration.minutes += extra_full_minutes as u32;
+                                    let extra_seconds =
+                                        ((extra_minutes - extra_full_minutes) * 60.0).round();
+                                    let extra_full_seconds = extra_seconds.trunc();
+                                    duration.seconds += extra_full_seconds as u32;
+                                    let micro_extra = ((extra_seconds - extra_full_seconds)
+                                        * 1_000_000.0)
+                                        .round()
+                                        as u32;
+                                    duration.microseconds += micro_extra;
+                                }
+                            }
+                            _ => {
+                                return Err(
+                                    self.parse_error("Invalid duration time unit".to_string())
+                                )
+                            }
+                        }
+                    }
+                }
+            }
+            self.inc();
+
+            if self.end() {
+                break;
+            }
+        }
+
+        parsed.duration = Some(duration);
+
+        Ok(())
+    }
+
+    fn parse_duration_number_frac(&mut self) -> Result<(u32, Option<f64>), ParseError> {
+        let value = self.parse_duration_number()?;
+        if self.current == '.' || self.current == ',' {
+            let mut decimal = 0_f64;
+            let mut denominator = 1_f64;
+            loop {
+                self.inc();
+
+                match self.current {
+                    c if c.is_ascii_digit() => {
+                        decimal *= 10.0;
+                        decimal += f64::from(c.to_digit(10).unwrap());
+                        denominator *= 10.0;
+                    }
+                    _ => return Ok((value, Some(decimal / denominator))),
+                }
+            }
+        } else {
+            Ok((value, None))
+        }
+    }
+
+    fn parse_duration_number(&mut self) -> Result<u32, ParseError> {
+        let mut value = match self.current {
+            c if c.is_ascii_digit() => c.to_digit(10).unwrap(),
+            _ => {
+                return Err(self.parse_error("Invalid number in duration".to_string()));
+            }
+        };
+
+        loop {
+            self.inc();
+
+            match self.current {
+                c if c.is_ascii_digit() => {
+                    value *= 10;
+                    value += c.to_digit(10).unwrap();
+                }
+                _ => return Ok(value),
+            }
+        }
```

Comment:
Use the `self.inc()` change from line 127 to simplify the looping condition.
```suggestion
        while let Some(digit) = self.inc().and_then(|ch| ch.to_digit(10)) {
            value *= 10;
            value += digit;
        }

        Ok(value)
```